### DISCUSSION TITLE
A question about snopt

### DISCUSSION THREAD

**Discussion by 101709** [mdolab/pyoptsparse#418] [General]
![屏幕截图 2024-11-30 160530](https://github.com/user-attachments/assets/fa98c42c-7d82-481e-94f5-39195ed5a8d0)
![屏幕截图 2024-11-30 160550](https://github.com/user-attachments/assets/a7cdce3d-4e91-4566-9194-bbb660ef9c4b)
As the picture shows, this is where I limit 'CD' during optimization to make 'CL' maximum, so CL's scale is set to -1. But I ran into a problem. My 'CL' iteration process, as shown in the figure, did not show the value of the snopt output file. For example, the four data of snopt output file: -7.7549929E-01, -8.5897671E-01, -8.6909302E-01, -8.5723396E-01, only the initial data is available, and the others do not appear in the CFD calculation process. Why is this? Is there something wrong with this output data

**Comment by 101709** [comment]
[opt_SNOPT_summary.txt](https://github.com/user-attachments/files/17964790/opt_SNOPT_summary.txt)
[opt_SNOPT_print.txt](https://github.com/user-attachments/files/17964791/opt_SNOPT_print.txt)
This is the entire output file

**Reply by 101709** [reply]
The problem I have is that the values of the MeritFunction and the objective function are very different. And the MeritFunction goes down, but the objective function goes up

**Comment by ewu63** [comment]
1. You're looking at the Merit Function which is not the same as the objective function value. Check the SNOPT manual for the definition.
2. The history contains all the finite difference steps you took, so the iterations will not match with SNOPT.
3. The objective function value increases because you are maximizing the CL, SNOPT sees the negative of the CL hence the negative values.

**Reply by 101709** [reply]
I may not be clear, I want to know how to reduce the gap between the MeritFunction and the objective function. Second, I do want to reduce the negative of CL, which means increasing CL. But now the problem is that the negative value of CL may be bigger than the initial value, that is, CL is smaller than it was before. Is there any way to fix this? Because I used 'ipopt' to optimize it so well. Thank you!This is my setup:
if args.opt == "snopt":
    optOptions = {
        "Major feasibility tolerance": 1.0e-6,
        "Major optimality tolerance": 1.0e-3,
        "Minor feasibility tolerance": 1.0e-6,
        "Verify level": -1,
        "Function precision": 1.0e-6,
        "Major iterations limit": 50,
        "Nonderivative linesearch": None,
        "Print file": "opt_SNOPT_print.txt",
        "Summary file": "opt_SNOPT_summary.txt",
    }
elif args.opt == "ipopt":
    optOptions = {
        "tol": 1.0e-3,
        "constr_viol_tol": 1.0e-6,
        "max_iter": 50,
        "print_level": 5,
        "output_file": "opt_IPOPT.txt",
        "mu_strategy": "adaptive",
        "limited_memory_max_history": 10,
        "nlp_scaling_method": "none",
        "alpha_for_y": "full",
        "recalc_y": "yes",
    }
